src/ClientWidgetApi.ts

```diff
@@ -514,21 +518,38 @@ export class ClientWidgetApi extends EventEmitter {
                 });
             }
         }
-
-        // For backwards compatibility we still call the deprecated
-        // readRoomEvents and readStateEvents methods in case the client isn't
-        // letting us know the currently viewed room via setViewedRoomId
-        const events = request.data.room_ids === undefined && askRoomIds.length === 0
-            ? await (
+        
+        let events: IRoomEvent[];
+
+        if(request.data.room_ids === undefined && askRoomIds.length === 0){
+            // For backwards compatibility we still call the deprecated
+            // readRoomEvents and readStateEvents methods in case the client isn't
+            // letting us know the currently viewed room via setViewedRoomId
+            events = await (
+                // This returns [] with the current driver of Element Web.
+                // Add default implementations of the `readRoomEvents` and `readStateEvents`
+                // methods to use `readRoomTimeline` and `readRoomState` if they are not overwritten.
```

Comment:
I thought it would make sense to provide default implementations of `readRoomTimeline` + `readRoomState` that fall back to the deprecated `readRoomEvents` and `readStateEvents` methods, for compatibility. But if we additionally give `readRoomEvents` and `readStateEvents` default implementations, and the client's driver implements none of these methods, then we have a circular dependency which will crash the client.

I believe it's okay to assume that if the client doesn't call `setViewedRoomId`, then it won't implement `readRoomTimeline` or `readRoomState` either. `setViewedRoomId` is basically a prerequisite for using the new driver methods. In other words, I don't see this as a case we need to be concerned with?

Reply:
While creating: https://github.com/element-hq/element-web/pull/29109
I also started to wonder if we can even hit that case. (If I am seeing this right in EW it is not possible since EW will never reach this path with `askRoomIds.length === 0` being true.)
So this case can only happen for clients implementing a widget driver with the matrix-widget api and implementing the new methods but failing to call setViewedRoom.

Is that correct?
I added a warning console log in case someone still does that.

Reply:
That matches my understanding, yeah.